### PR TITLE
Remove hardcoded translations from Vue

### PR DESCRIPTION
**Pull request by christophehenry**
- [x] Use VueRouter to match the route and render different stamped elements
- [x] Use Django's gettext library to perform translations
- [x] I can haz tests?

### CODE REVIEW

File: socialhome/streams/app/components/Stream.vue [hunk 1]
```diff
@@ -63,6 +62,9 @@ export default Vue.component("stream", {
     beforeDestroy() {
         this.$store.$websocket.close()
     },
+    created(){
+        console.log(this.$route.path)
+    }
```
Comment (christophehenry) [inline socialhome/streams/app/components/Stream.vue]: Nope!

File: socialhome/streams/app/components/FollowedStampedElement.vue [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import GenericStampedElement from "streams/app/components/GenericStampedElement.vue"
+
+
+export default Vue.component("followed-stamped-element", {
+    computed: {
+        title() {
+            return gettext("Followed")
+        },
+        helpText() {
+            return gettext("Contains public content from around the network.")
```
Comment (christophehenry) [inline socialhome/streams/app/components/FollowedStampedElement.vue]: Should be `"Content from followed users."`

**Comment by jaywink** [comment]
With this solution, one general concern I have is how streams mapped to root `/` will be handled. Can `vue-router` handle complex "if/else" logic for a single route? If it can, we *could* export the user preferences controlling what is rendered and then use those preferences to decide like in Django [here](https://github.com/jaywink/socialhome/blob/f048f0adc5c07ce48e562351ae78dabf799a2622/socialhome/views.py#L16).

But to be honest, you will need to base what stream is rendered (or what to fetch via API) anyway based on the stream name - or the exported preferences.

To me, it feels it would be simpler just looking at "what is stream name" and then rendering the right stamped element component based on that. It's much simpler and the decision which sub component to render is in the right place, the `StreamComponent`. Doing this way feels like a bit of a hack to use the router to decide on which sub component is rendered. I understand the case for a router if we actually *do routing* inside Vue. But this is more of a "if route this, render this component  when rendering the stream, which isn't itself routed".

So, TL/DR, I support the idea of different components for the different stream stamped elements (a must for profile stamped element, it's quite complex), but I feel there is a simpler way to decide which one to render. A stream is *not* mapped to a single URL in Socialhome, so you can't absolutely decide that Followed stream is always `/followed`.

**Comment by christophehenry** [comment]
With this solution, one general concern I have is how streams mapped to root `/` will be handled. Can `vue-router` handle complex "if/else" logic for a single route? If it can, we could export the user preferences controlling what is rendered and then use those preferences to decide like in Django [here](https://github.com/jaywink/socialhome/blob/f048f0adc5c07ce48e562351ae78dabf799a2622/socialhome/views.py#L16).

I could specify multiple components [here](https://github.com/jaywink/socialhome/pull/281/files#diff-14c22c262c80bfa58a8ccc08463d9ae2R10) and change [the name](https://github.com/jaywink/socialhome/pull/281/files#diff-f6bf769c688863e153cb8f8d589a3803R13) using a data property in the component.

> To me, it feels it would be simpler just looking at "what is stream name" and then rendering the right stamped element component based on that.

When the time of vueifying profiles pages come, I don't want to do things [like that](https://github.com/jaywink/socialhome/blob/master/socialhome/streams/app/stores/streamStore.state.js#L18). Having a stream name variable feels wrong anyway.

>  I understand the case for a router if we actually _do routing_ inside Vue.

Oh, we will do routing very soon. I intend to render any stream on front-end anyway so we can use caching of data client-side and epically reduce network load.

File: socialhome/streams/app/components/GenericStampedElement.vue [hunk 3]
```diff
@@ -15,5 +15,11 @@ import Vue from "vue"
 import store from "streams/app/stores/applicationStore"
 
 
-export default Vue.component("stamped-element", {store})
+export default Vue.component("generic-stamped-element", {
+    store,
```
Comment (christophehenry) [inline socialhome/streams/app/components/GenericStampedElement.vue]: Useless, now

File: socialhome/streams/app/components/FollowedStampedElement.vue [hunk 4]
```diff
@@ -0,0 +1,20 @@
+<template>
+    <generic-stamped-element :title="title" :help-text="helpText" />
+</template>
+
+<script>
+import Vue from "vue"
+import GenericStampedElement from "streams/app/components/GenericStampedElement.vue"
```
Comment (jaywink) [inline socialhome/streams/app/components/FollowedStampedElement.vue]: `Generic` kind of keeps colliding with me every time I look at this. Can we use `BaseStampedElement`? "Base" is a more common word to use in these cases for example with Django templates or classes.

File: socialhome/streams/app/components/Stream.vue [hunk 5]
```diff
@@ -52,6 +53,18 @@ export default Vue.component("stream", {
             },
         }
     },
+    computed: {
+        stampedElement() {
```
Comment (jaywink) [inline socialhome/streams/app/components/Stream.vue:59]: A simple test for this computed function would be nice 👍

**Review by jaywink — COMMENTED** [review]
Looks awesome now! A few comments.

**Comment by christophehenry** [comment]
Done

File: socialhome/streams/app/components/Stream.vue [hunk 6]
```diff
@@ -52,6 +55,19 @@ export default Vue.component("stream", {
...
+            if (this.$store.state.streamName.match(/followed/)) {
```
Comment (jaywink) [inline socialhome/streams/app/components/Stream.vue]: You must use `starts with` kind of check OR do a split and take the first element. For example currently tag streams have the tag name slug in the stream name, so #followed would get the followed stamped element. In a way this is bad and unnecessary, but it's the way it is. And the "spec" for stream names is basically `<stream_type>__<id or other identifier>`. The first two underscores end the stream type definition which should be used.

File: socialhome/streams/app/components/StreamElement.vue [hunk 7]
```diff
@@ -1,13 +1,17 @@
 <template>
     <div>
-        <div v-html="htmlSafe"></div>
+        <div v-if="nsfw" class="card card-block text-center nsfw-shield">
+            <b-button href="#" variant="link" @click.stop.prevent="toggleNsfw">
+                {{ toggleNsfwBtnText }}
+            </b-button>
+        </div>
+        <div v-if="showNsfw" v-html="htmlSafe"></div>
```
Comment (jaywink) [inline socialhome/streams/app/components/StreamElement.vue]: This is incorrect. The NSFW shield is already in the rendered contents in Socialhome. There is an `nsfw` class added to all images in the content if a tag is found. This activates the NSFW shield. What you need in the JS side is a trigger to show it on click. There should also be a trigger then to return the nsfw shield. Try it to see how it works.

We don't want to hide the whole content, just images. Diaspora NSFW shield always annoyed me because you can't see any context to the post before clicking it.

File: socialhome/streams/app/components/TagStampedElement.vue [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+import "streams/app/components/BaseStampedElement.vue"
+
+
+export default Vue.component("tag-stamped-element", {
+    computed: {
+        title() {
+            return `#${this.$store.state.tagName}`
+        },
+        helpText() {
+            return interpolate(gettext("All content tagged %s"), [`#${this.$store.state.tagName}`])
```
Comment (jaywink) [inline socialhome/streams/app/components/TagStampedElement.vue]: Used to be just the tag name, but if you want to put some more text that is fine. Make sure it ends in a period though ;) Also, I think it should be "All content tagged *with* %s".

File: socialhome/streams/app/tests/components/Stream.tests.js [hunk 9]
```diff
@@ -14,6 +16,28 @@ describe("Stream", () => {
...
+                let target = mount(Stream, {})
+                target.instance().$store.state.streamName = "followed"
+                target.instance().stampedElement.should.eq("FollowedStampedElement")
+            })
+            it("should render the PublicStampedElement when stream name is 'public'", () => {
+                let target = mount(Stream, {})
+                target.instance().$store.state.streamName = "public"
+                target.instance().stampedElement.should.eq("PublicStampedElement")
+            })
+            it("should display an error when stream name is unknown", () => {
```
Comment (jaywink) [inline socialhome/streams/app/tests/components/Stream.tests.js]: Missing test for tag stream.

File: socialhome/streams/app/tests/components/Stream.tests.js [hunk 10]
```diff
@@ -14,6 +16,28 @@ describe("Stream", () => {
...
+                let target = mount(Stream, {})
+                Sinon.spy(console, "error")
+                target.instance().$store.state.streamName = "Yolo stream"
```
Comment (jaywink) [inline socialhome/streams/app/tests/components/Stream.tests.js]: Should we make a Yolo stream??

**Review by jaywink — CHANGES_REQUESTED** [review]
Unfortunately the NSFW behaviour does not mirror what exists now.

File: socialhome/streams/app/components/Stream.vue (hunk 6 above)
Comment (christophehenry) [reply socialhome/streams/app/components/Stream.vue]: `/^followed/` might be enough, then.

File: socialhome/streams/app/components/StreamElement.vue (hunk 7 above)
Comment (christophehenry) [reply socialhome/streams/app/components/StreamElement.vue]: Hmm... I don't like that. It's not the Vue way. It will require messing with selectors :/

Comment (jaywink) [reply socialhome/streams/app/components/StreamElement.vue]: Can't bend features to code.. Code must bend to features ;)

File: socialhome/streams/app/tests/components/Stream.tests.js (hunk 10 above)
Comment (christophehenry) [reply socialhome/streams/app/tests/components/Stream.tests.js]: Most definitely!

**Comment by christophehenry** [comment]
Ready for another review

File: socialhome/streams/app/components/TagStampedElement.vue [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+import "streams/app/components/BaseStampedElement.vue"
+
+
+export default Vue.component("tag-stamped-element", {
+    computed: {
+        title() {
+            return `#${this.$store.state.tagName}`
+        },
+        helpText() {
+            return interpolate(gettext("All content tagged with %s"), [`#${this.$store.state.tagName}`])
```
Comment (jaywink) [inline socialhome/streams/app/components/TagStampedElement.vue]: Missing period from sentence 👍

**Review by jaywink — CHANGES_REQUESTED** [review]
One missing period and then merge! 💃

**Review by jaywink — APPROVED** [review]

**Comment by jaywink** [comment]
Thanks!
